File: rov_modes.py
"""Sumber kebenaran untuk pilot mode ArduSub di sisi Python.

Nama GUI:
    manual
    stabilize
    depth_hold
    poshold

Nama ArduSub:
    MANUAL
    STABILIZE
    ALT_HOLD

Depth Hold menggunakan ALT_HOLD ArduSub secara langsung.
Tidak ada lagi depth-set atau depth-bias custom di Python.

POSHOLD adalah overlay heading-hold sisi Python yang berjalan
di atas ALT_HOLD ArduSub. Overlay ini menahan heading, bukan
posisi horizontal.
"""

# =========================
# Pemetaan pilot mode
# =========================

PILOT_MODE_MAP = {
    "manual": "MANUAL",
    "stabilize": "STABILIZE",
    "depth_hold": "ALT_HOLD",
    "poshold": "ALT_HOLD",
}

# Nama GUI untuk station-keep.
POSHOLD_MODE = "poshold"

# Mode ArduSub tempat overlay POSHOLD berjalan.
# POSHOLD firmware tidak digunakan karena ROV tidak mempunyai
# sumber posisi horizontal yang diperlukan.
POSHOLD_BASE_MODE = PILOT_MODE_MAP[POSHOLD_MODE]


def resolve_pilot_mode(name):
    """Terjemahkan nama mode GUI ke nama mode ArduSub.

    Menerima huruf besar/kecil dan spasi di tepi.
    Mengembalikan None jika nama mode tidak dikenal.
    """
    if not isinstance(name, str):
        return None

    return PILOT_MODE_MAP.get(name.strip().lower())


def is_poshold_request(name):
    """True jika nama mode GUI meminta overlay POSHOLD."""
    if not isinstance(name, str):
        return False

    return name.strip().lower() == POSHOLD_MODE


def poshold_mode_ok(mode):
    """True jika overlay heading-hold boleh bekerja pada mode tersebut."""
    return mode == POSHOLD_BASE_MODE

File: shared/rov-modes.js
/* Nama & gating pilot mode ArduSub — SATU sumber kebenaran sisi JS, dipakai
 * baik oleh browser (public/js/app.js) maupun test Node (server/test/*).
 *
 * Padanan sisi Python: rov_modes.py. Jaga keduanya tetap sinkron kalau ada
 * mode baru — tidak ada mekanisme otomatis yang mengecek ini lintas bahasa.
 */

/* data-mode tab GUI / value command pilot_mode -> nama mode ArduSub (HEARTBEAT).
 *
 * Alasan lengkap tiap pemetaan ada di docstring rov_modes.py. */
export const PILOT_MODE_MAP = {
  manual: "MANUAL",
  stabilize: "STABILIZE",
  depth_hold: "ALT_HOLD",
  /* "poshold" adalah OVERLAY sisi Pi di atas ALT_HOLD (tahan kedalaman + tahan
   * heading), BUKAN mode POSHOLD firmware ArduSub — itu butuh estimasi posisi
   * horizontal dari EKF yang tidak ada di bawah air. Ia TIDAK menahan posisi
   * x/y. Alasan lengkapnya di docstring rov_modes.py. */
  poshold: "ALT_HOLD",
};

/* Nama mode ArduSub (dari HEARTBEAT) -> data-mode tab GUI.
 *
 * "poshold" TIDAK ada di sini, dan itu disengaja: peta ini satu-arah,
 * sedangkan ALT_HOLD bisa berarti dua tab (Alt Hold atau Pos Hold). Yang
 * memisahkan keduanya adalah flag `poshold` di telemetri — lihat syncModeTabs()
 * di app.js. */
export const ARDUSUB_MODE_TO_TAB = {
  MANUAL: "manual",
  STABILIZE: "stabilize",
  ALT_HOLD: "depth_hold",
};

/* Mode yang jadi syarat bias depth-set, sehingga bias throttle punya sesuatu
 * untuk dikoreksi (atau, di STABILIZE, jadi satu-satunya yang mendorong).
 * Padanan rov_modes.DEPTH_HOLD_MODES.
 *
 * BUKAN saklar depth-set: berada di STABILIZE tidak menyalakan depth-set. GUI
 * memakainya hanya untuk membedakan label ARMED (saklar hidup, mode belum
 * benar) dari HOLDING (bias benar-benar mengalir). */
export const DEPTH_HOLD_MODES = new Set(["STABILIZE"]);
